### PR TITLE
Add filter method

### PR DESCRIPTION
**Pull request by jbasrai**
Scala's Option has a [filter](https://www.scala-lang.org/api/current/scala/Option.html#filter(p:A=>Boolean):Option[A]) method that is useful. I've found myself shimming this method for this library, so thought it might be worth including directly.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -57,6 +60,12 @@ Some.prototype.flatMap = function(func) {
     return func(this._value);
 };
 
+Some.prototype.filter = function(f) {
+    return this.flatMap(function(x) {
+        return f(x) ? exports.some(x) : exports.none;
+    });
```
Comment (mwilliamson) [inline index.js]: This seems a little overcomplicated. Wouldn't the following accomplish the same thing?

```javascript
Some.prototype.filter = function(predicate) {
    return predicate(this._value) ? this : none;
};
```

File: test/options.test.js [hunk 2]
```diff
@@ -141,3 +141,34 @@ exports.callingValueOnSomeReturnsValue = function(test) {
...
+exports.filteringNoneReturnsNone = function(test) {
+    function equals3(x) {
+        return x === 3;
+    }
+
+    test.deepEqual(options.none.filter(equals3), options.none);
+    test.done();
+};
+
+exports.filteringSomeReturnsIdentityIfPredicateIsTrue = function(test) {
```
Comment (mwilliamson) [inline test/options.test.js]: This test name doesn't seem quite right to me -- filtering when the predicate is true doesn't return the identity, it's equivalent to applying the identity. I think it'd be clearer to write the test name (including the other tests for filter) similarly to the docs e.g.

```javascript
exports["given predicate(value) is true then some(value).filter(predicate) is some(value)"] = ...
```

What do you think?

File: README.md [hunk 3]
```diff
@@ -48,6 +48,12 @@ Conventionally used when `func` returns another option.
 * `some(value).flatMap(func)` returns `func(value)`
 * `none.flatMap(func)` returns `none`
 
+### filter(func)
```
Comment (mwilliamson) [inline README.md]: Minor thing, but `func` should be surrounded by asterisks. Also, for `filter`, would calling the argument `predicate` be clearer than `func`?

File: index.js [hunk 4]
```diff
@@ -57,6 +60,12 @@ Some.prototype.flatMap = function(func) {
     return func(this._value);
 };
 
+Some.prototype.filter = function(f) {
```
Comment (mwilliamson) [inline index.js]: Would calling the argument `predicate` be clearer than `f`?

**Review by mwilliamson — CHANGES_REQUESTED** [review]
Thanks for the pull request, adding a `filter` method sounds sensible to me. I've left some comments on some of the code, let me know if anything doesn't make sense or you think I'm talking nonsense.

File: README.md (hunk 3 above)
Comment (jbasrai) [reply README.md]: Yup agree. I've changed all `func` arguments to be `predicate` instead.

File: index.js (hunk 4 above)
Comment (jbasrai) [reply index.js]: Yes!

File: index.js (hunk 1 above)
Comment (jbasrai) [reply index.js]: Totally, this implementation is so much cleaner.

File: test/options.test.js (hunk 2 above)
Comment (jbasrai) [reply test/options.test.js]: Agree! Think I'm using that word without knowing what it actually means. I've changed the names of the tests to be more practical similar to what you suggested.

**Review by jbasrai — COMMENTED** [review]
Thanks for the feedback. Completely agree with all the points you've made. I've made changes to reflect everything discussed.

**Comment by jbasrai** [comment]
Sorry if I'm missing something in the review process. I noticed that the "Changes requested" label is still up. Let me know if I need to do something on my end besides pushing the changes!

**Comment by mwilliamson** [comment]
Looks good, thanks!

**Pull request merged**